Introduce runnerParentFirstArtifact for use with RunnerClassLoader

geoand opened the pull request:
Fixes: #13216

geoand committed:
Introduce runnerParentFirstArtifact for use with RunnerClassLoader

Fixes: #13216

geoand commented:
@stuartwdouglas @Sanne this obviously isn't the proper solution, just wanted to show what artifacts must absolutely be loaded by the JDK ClassLoader.

Perhaps we need another layer of modelling in the parentFirst artifact declaration

geoand commented:
A simple solution here would be to introduce another configuration knob, the `runnerParentFirst` and add the artifacts I hard coded in this PR. 
WDYT?

geoand commented:
I updated the PR to reflect my comment above.

dmlloyd commented:
I don't think we want to make this configurable unless absolutely necessary...

dmlloyd commented on core/runtime/pom.xml:
```diff
@@ -150,6 +150,19 @@
                         <parentFirstArtifact>org.junit.platform:junit-platform-commons</parentFirstArtifact>
                         <parentFirstArtifact>org.junit.platform:junit-platform-engine</parentFirstArtifact>
                     </parentFirstArtifacts>
+                    <runnerParentFirstArtifacts>
+                        <runnerParentFirstArtifact>io.quarkus:quarkus-bootstrap-runner</runnerParentFirstArtifact>
+                        <runnerParentFirstArtifact>io.quarkus:quarkus-development-mode-spi</runnerParentFirstArtifact>
+                        <!-- logging dependencies always need to be loaded by the JDK ClassLoader -->
+                        <runnerParentFirstArtifact>org.jboss.logging:jboss-logging</runnerParentFirstArtifact>
+                        <runnerParentFirstArtifact>org.jboss.logmanager:jboss-logmanager-embedded</runnerParentFirstArtifact>
+                        <runnerParentFirstArtifact>org.jboss.slf4j:slf4j-jboss-logmanager</runnerParentFirstArtifact>
+                        <runnerParentFirstArtifact>org.slf4j:slf4j-api</runnerParentFirstArtifact>
+                        <!-- The following dependencies are dependencies of the JBoss Logging so they too need  be loaded by the JDK ClassLoader -->
+                        <runnerParentFirstArtifact>io.smallrye.common:smallrye-common-io</runnerParentFirstArtifact>
+                        <runnerParentFirstArtifact>org.wildfly.common:wildfly-common</runnerParentFirstArtifact>
+                        <runnerParentFirstArtifact>org.graalvm.sdk:graal-sdk</runnerParentFirstArtifact>
+                    </runnerParentFirstArtifacts>
```
Are you sure about all of these?  In particular the SLF4J ones shouldn't be necessary, nor should the SmallRye one.  Also the `svm` one shouldn't be needed either since it's only used for annotations AFAICT.

geoand replied:
Right, the slf4j stuff isn't necessary.

The smallrye stuff is necessary however (see the updated comments)

dmlloyd replied:
What about the graalvm SDK?  What needs that at run time?

geoand replied:
That one is needed because of: https://github.com/quarkusio/quarkus/blob/966b5045e8f874060be06c6c7aaa0a074a93bd9b/independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/logging/InitialConfigurator.java#L51

dmlloyd replied:
Ah I see.  I wonder if that's worth replacing it with the contents of that method:

```java
        if (System.getProperty("org.graalvm.nativeimage.imagecode").equals("runtime")) {
```

geoand replied:
Good idea! Let me do that

geoand replied:
I pushed a commit that removes the dependency on the GraalVM jar

gsmet replied:
Hmmm. What does it bring to avoid having it there?

I'm not sure it's such a good idea to not rely on proper APIs.

dmlloyd replied:
Just minimizing the class path size, particularly with a JAR that is never really used.  Note that the system property name and value are both stable APIs as well.

dmlloyd commented:
I think we should hard code the ones we know will always be needed.  If absolutely necessary, we could let the user add more with some additional config, but this would be a very unusual situation I would think, right?

geoand commented:
> I think we should hard code the ones we know will always be needed. If absolutely necessary, we could let the user add more with some additional config, but this would be a very unusual situation I would think, right?

Yeah, it would absolutely be very unusual - I seriously doubt we would ever need to do it.

The reason I went with the XML config in the `quarkus-core`'s `pom.xml` file is that it's in line with what we have done so far and therefore much easier for us to find in a few months we have forgot all about where all this is configured - as opposed to having to dig through the source and find some other way configuration.

I don't have a strong opinion on this however. 
I'd like to know what Stuart thinks as well.

dmlloyd commented:
As long as it's not something that users can mess around with (i.e. it only affects our build) then it's fine.  I suppose that requiring the user to create their own custom Quarkus build is a high enough bar to prevent them from doing something foolish...

geoand committed:
Replace org.graalvm.sdk:graal-sdk dependency in runner bootstrap with custom class

This is done in order to avoid making this dependency a runtimeParentDependency,
which means that we can reduce the number of jars loaded by the JDK ClassLoader

geoand commented:
> As long as it's not something that users can mess around with (i.e. it only affects our build) then it's fine. I suppose that requiring the user to create their own custom Quarkus build is a high enough bar to prevent them from doing something foolish...

Very true. If a user wanted to fiddle with this, then yes they would need to create their own extension. If someone is willing to go to that length, then :shrug: :)

Sanne commented:
Nice! I tried this with another heap dump, still found these references in memory:

```
target/quarkus-app/quarkus-run.jar
target/quarkus-app/lib/boot/io.quarkus.quarkus-development-mode-spi-999-SNAPSHOT.jar
target/quarkus-app/lib/boot/org.jboss.logging.jboss-logging-3.4.1.Final.jar
target/quarkus-app/lib/boot/org.jboss.logmanager.jboss-logmanager-embedded-1.0.6.jar
target/quarkus-app/lib/boot/org.wildfly.common.wildfly-common-1.5.4.Final-format-001.jar
target/quarkus-app/lib/boot/io.smallrye.common.smallrye-common-io-1.4.1.jar
target/quarkus-app/lib/boot/io.quarkus.quarkus-bootstrap-runner-999-SNAPSHOT.jar
target/quarkus-app/app/hibernate-orm-quickstart-1.0-SNAPSHOT.jar
target/quarkus-app/lib/main/org.jboss.threads.jboss-threads-3.1.1.Final.jar
target/quarkus-app/lib/main/org.jboss.narayana.jta.narayana-jta-5.10.6.Final.jar
```

the last two seem to have been reloaded after the caches had been invalidated; I can try getting a stacktrace from those to check if it's avoidable.

geoand commented:
> 
> the last two seem to have been reloaded after the caches had been invalidated; I can try getting a stacktrace from those to check if it's avoidable.

That would be nice, thanks

Sanne commented:
> > the last two seem to have been reloaded after the caches had been invalidated; I can try getting a stacktrace from those to check if it's avoidable.
> 
> That would be nice, thanks

To be clear, I believe it's unavoidable that some libraries will need to re-open the jars. I want to check these out so to learn more about it - and possibly see if we should have test/diagnostics helpers to make sure such cases are "approved" rather than unintentional.
But don't consider it a blocker for your fix, it's unrelated.

geoand commented:
Yes I know. But still since you have done all this analysis, it would nice to know the final piece even though it doesn't affect this fix

Sanne commented:
so it turns out Vertx does something I didn't expect, and a single vertx ServiceLoader initialization is having us reload all jars.

Vertx has a caching layer implemented by `io.vertx.core.file.impl.FileResolver` which triggers a full-scan for all `META-INF/resources`; when this is combined by Quarkus's `RunnerClassLoader` which is mapping package/jar, this triggers reloading of most jar files - as this path is commonly used among all of them.

I couldn't look much into this, but we'll need to either bypass that loading strategy or have it fill its caches in advance. And loading of `META-INF/resources` is always going to be a problem, maybe we need a smarter plan for loaders of such resources - for example, I'd love it if we could use the explicit resource & service registration build items to be enforced for JVM mode as well.

dmlloyd commented:
Another possible strategy could be for the class loader itself to preload those (and possibly other) resources and keep them in memory.

stuartwdouglas commented:
So after the close you could default to 'open on demand' for the zip file, but keep a count. If this count his a large enough number (say 10,000 for the sake of the argument) then you change back to open by default. 

It is still perfectly normal for classes to be loaded after the app has started, and this could happen quite a while after startup (e.g. if there is some business logic that is only executed at night, this could load lots of related classes). 

This counter approach means that late loaded classes won't cause the zip files to end up in memory, but things like web resources being served in a benchmark will hit the counter and trigger the zip to stay open. 

We could look at other approaches for web resources as well, such as special casing the handling of META-INF/resources and storing them unzipped on the file system rather than serving them from a jar.

geoand commented:
@stuartwdouglas are you OK with the approach taken in this PR to configure the RunnerClassLoader?

Sanne commented:
that's a lot of most excellent suggestions! I'll break it up in a separate issue so that the brainstorming doesn't get in the way of feedback for the current PR.

For resources loading: #13256 
For deferred Jar closing: #13258

stuartwdouglas reviewed (changes requested):
LGTM but the sentry extension will also need to be updated.

geoand commented:
Right!

PR Updated
